apps/admin_app/assets/js/components/CreatePromotions.tsx

```diff
@@ -0,0 +1,369 @@
+            selectedCalculator:""
+        }
+
+        this.handleStartDate = this.handleStartDate.bind(this);
+        this.handleEndDate = this.handleEndDate.bind(this);
+        this.handleSave = this.handleSave.bind(this);
+        this.handleAddRule = this.handleAddRule.bind(this); 
+        this.fetchPOST = this.fetchPOST.bind(this);
+        this.handleAddAction = this.handleAddAction.bind(this);
+        this.handleRule = this.handleRule.bind(this);
+        this.renderRuleOptions = this.renderRuleOptions.bind(this);
+        this.handleCalculator = this.handleCalculator.bind(this);
+        this.renderCalculatorOptions = this.renderCalculatorOptions.bind(this);
+        this.handleSaveRule = this.handleSaveRule.bind(this);
+        this.onSuccess = this.onSuccess.bind(this);
+        this.handleDeleteRule = this.handleDeleteRule.bind(this);
+    }
+
+    handleStartDate = (date,dateString)=>{
+        const startDate = new Date(dateString)
+        startDate.setHours(0,0,0)
+        this.setState({starts_at:startDate})
+    }
+
+    handleEndDate = (date,dateString)=>{
+        const endDate = new Date(dateString)
+        endDate.setHours(0,0,0)
+        this.setState({expires_at:endDate})
+    }
+
+    fetchPOST = (url,data)=>{
+        return fetch(url,{
+            credentials: 'include',
+            method:'POST',
+            body:JSON.stringify(data),
+            headers: {'Content-Type': 'application/json; charset=UTF-8'},
+        }).then(res =>res.json())
+        
+    }
+
+    fetchGET =(url)=>{
+        return fetch(url,{
+            credentials:'include',
+            method:"GET",
+            headers:{'Content-Type': 'application/json; charset=UTF-8'}
+        }).then(res =>res.json())
+    }
+
+    handleSave = () =>{
+        var data = {
+            "data":{
+                "starta_at":this.state.starts_at,
+                "expires_at":this.state.expires_at,
+                "code":this.state.code,
+                "name":this.state.name,
+                "rules":[],
+                "actions":[]
+            }
+        }
+        var url = 'http://localhost:4000/api/promotions'
+        this.fetchPOST(url,data).then(response=>{
+            var arrofkeys=Object.keys(response)
+            arrofkeys[0]=="errors"?this.onFail(response):this.onSuccess()
+            console.log('Success',JSON.stringify(response))})
+        .catch(error=>console.error('Error',error));
+        
+    }
+
+    onSuccess =()=>{
+        this.setState({success:true})
+        alert("Saved Successfully!")
+    }
+    onFail=(response)=>{
+        Object.keys(response).map((res)=>{
+            var errormsg=""
+            Object.keys(response[res]).map((res1)=>{
+                errormsg=errormsg+res1+" "+response[res][res1][0]["message"]+"\n"
+            })
+            alert(errormsg)
+        })
+    }
+    handleAddRule=()=>{
+        
+        this.setState({addRule:true})
+        var url="http://localhost:4000/api/promo-rules"
```

Comment:
using var instead of const and hardcoded url - Not ideal way.
Put this url inside some constants file and then use here.